In [92]:
import psycopg2
import pandas as pd


In [93]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [94]:
connection = psycopg2.connect(
   dbname=DBNAME,
   user=USER,
   host=HOST,
   password=PASSWORD,
   port=PORT
)

In [95]:
# Задание 3.1. Сколько вакансий есть в базе?
query_3_1 = """
SELECT COUNT(*) as total_vacancies
FROM vacancies;
"""

result_3_1 = pd.read_sql_query(query_3_1, connection)
print("3.1. Количество вакансий в базе:")
print(result_3_1)
print(f"Ответ: {result_3_1.iloc[0]['total_vacancies']} вакансий")


3.1. Количество вакансий в базе:
   total_vacancies
0            49197
Ответ: 49197 вакансий


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\1121738544.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_3_1 = pd.read_sql_query(query_3_1, connection)


In [96]:
# Задание 3.2. Сколько работодателей?
query_3_2 = """
SELECT COUNT(*) as total_employers
FROM employers;
"""

result_3_2 = pd.read_sql_query(query_3_2, connection)
print("3.2. Количество работодателей:")
print(result_3_2)
print(f"Ответ: {result_3_2.iloc[0]['total_employers']} работодателей")


3.2. Количество работодателей:
   total_employers
0            23501
Ответ: 23501 работодателей


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\2803318637.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_3_2 = pd.read_sql_query(query_3_2, connection)


In [97]:
# Задание 3.3. Сколько регионов?
query_3_3 = """
SELECT COUNT(*) as total_areas
FROM areas;
"""

result_3_3 = pd.read_sql_query(query_3_3, connection)
print("3.3. Количество регионов:")
print(result_3_3)
print(f"Ответ: {result_3_3.iloc[0]['total_areas']} регионов")


3.3. Количество регионов:
   total_areas
0         1362
Ответ: 1362 регионов


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\995873020.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_3_3 = pd.read_sql_query(query_3_3, connection)


In [98]:
# Задание 3.4. Сколько сфер деятельности?
query_3_4 = """
SELECT COUNT(*) as total_industries
FROM industries;
"""

result_3_4 = pd.read_sql_query(query_3_4, connection)
print("3.4. Количество сфер деятельности:")
print(result_3_4)
print(f"Ответ: {result_3_4.iloc[0]['total_industries']} сфер деятельности")


3.4. Количество сфер деятельности:
   total_industries
0               294
Ответ: 294 сфер деятельности


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\4004968110.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_3_4 = pd.read_sql_query(query_3_4, connection)


In [99]:
# Задание 4.1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
query_4_1 = """
SELECT 
    a.name as area,
    COUNT(v.id) as cnt
FROM areas a
LEFT JOIN vacancies v ON a.id = v.area_id
GROUP BY a.id, a.name
ORDER BY cnt DESC;
"""

result_4_1 = pd.read_sql_query(query_4_1, connection)
print("4.1. Количество вакансий в каждом регионе:")
print(result_4_1.head(10))  # Показываем топ-10 регионов
print(f"\nВсего регионов с вакансиями: {len(result_4_1[result_4_1['cnt'] > 0])}")
print(f"Регионов без вакансий: {len(result_4_1[result_4_1['cnt'] == 0])}")


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\1119163137.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_4_1 = pd.read_sql_query(query_4_1, connection)


4.1. Количество вакансий в каждом регионе:
              area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892
5     Екатеринбург  1698
6  Нижний Новгород  1670
7           Казань  1415
8        Краснодар  1301
9           Самара  1144

Всего регионов с вакансиями: 769
Регионов без вакансий: 593


In [100]:
# Задание 4.2. У какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой?
query_4_2 = """
SELECT COUNT(*) as vacancies_with_salary
FROM vacancies
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL;
"""

result_4_2 = pd.read_sql_query(query_4_2, connection)
print("4.2. Количество вакансий с заполненной зарплатой:")
print(result_4_2)
print(f"Ответ: {result_4_2.iloc[0]['vacancies_with_salary']} вакансий")


4.2. Количество вакансий с заполненной зарплатой:
   vacancies_with_salary
0                  24073
Ответ: 24073 вакансий


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\3544661116.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_4_2 = pd.read_sql_query(query_4_2, connection)


In [101]:
# Задание 4.3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого числа.
query_4_3 = """
SELECT 
    ROUND(AVG(salary_from)) as avg_salary_from,
    ROUND(AVG(salary_to)) as avg_salary_to
FROM vacancies
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL;
"""

result_4_3 = pd.read_sql_query(query_7, connection)
print("4.3. Средние значения зарплатной вилки:")
print(result_4_3)
print(f"Ответ: Средняя нижняя граница: {result_4_3.iloc[0]['avg_salary_from']} руб.")
print(f"Средняя верхняя граница: {result_4_3.iloc[0]['avg_salary_to']} руб.")


4.3. Средние значения зарплатной вилки:
   avg_salary_from  avg_salary_to
0          71065.0       110537.0
Ответ: Средняя нижняя граница: 71065.0 руб.
Средняя верхняя граница: 110537.0 руб.


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\2499721922.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_4_3 = pd.read_sql_query(query_7, connection)


In [102]:
# Задание 4.4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Какая пара находится на втором месте по популярности?
query_4_4 = """
SELECT 
    schedule,
    employment,
    COUNT(*) as cnt
FROM vacancies
GROUP BY schedule, employment
ORDER BY cnt DESC;
"""

result_4_4 = pd.read_sql_query(query_4_4, connection)
print("4.4. Количество вакансий по сочетанию графика работы и типа трудоустройства:")
print(result_4_4)
print(f"\nВторая по популярности пара: {result_4_4.iloc[1]['schedule']} + {result_4_4.iloc[1]['employment']} ({result_4_4.iloc[1]['cnt']} вакансий)")


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\2096578324.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_4_4 = pd.read_sql_query(query_4_4, connection)


4.4. Количество вакансий по сочетанию графика работы и типа трудоустройства:
            schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа

In [103]:
# Задание 4.5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.
query_4_5 = """
SELECT 
    experience,
    COUNT(*) as cnt
FROM vacancies
GROUP BY experience
ORDER BY cnt ASC;
"""

result_4_5 = pd.read_sql_query(query_4_5, connection)
print("4.5. Требуемый опыт работы в порядке возрастания количества вакансий:")
print(result_4_5)


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\836887490.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_4_5 = pd.read_sql_query(query_4_5, connection)


4.5. Требуемый опыт работы в порядке возрастания количества вакансий:
           experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152


In [104]:
# Задание 5.1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.
query_5_1 = """
SELECT 
    e.name as employer_name,
    COUNT(v.id) as vacancy_count
FROM employers e
LEFT JOIN vacancies v ON e.id = v.employer_id
GROUP BY e.id, e.name
ORDER BY vacancy_count DESC
LIMIT 5;
"""

result_5_1 = pd.read_sql_query(query_5_1, connection)
print("5.1. Топ-5 работодателей по количеству вакансий:")
print(result_5_1)
print(f"\n1-е место: {result_5_1.iloc[0]['employer_name']} ({result_5_1.iloc[0]['vacancy_count']} вакансий)")
print(f"5-е место: {result_5_1.iloc[4]['employer_name']} ({result_5_1.iloc[4]['vacancy_count']} вакансий)")


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\3898817147.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_1 = pd.read_sql_query(query_5_1, connection)


5.1. Топ-5 работодателей по количеству вакансий:
   employer_name  vacancy_count
0         Яндекс           1933
1     Ростелеком            491
2       Тинькофф            444
3           СБЕР            428
4  Газпром нефть            331

1-е место: Яндекс (1933 вакансий)
5-е место: Газпром нефть (331 вакансий)


In [105]:
# Задание 5.2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём. Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.
query_5_2 = """
WITH employers_count_per_area AS (
    SELECT 
        area,
        COUNT(DISTINCT id) AS employers_count
    FROM employers
    GROUP BY area
),
vacancies_count_per_area AS (
    SELECT 
        area_id,
        COUNT(id) AS vacancies_count
    FROM vacancies
    GROUP BY area_id
)

SELECT 
    a.name AS area,
    COALESCE(ec.employers_count, 0) AS employers_count,
    COALESCE(vc.vacancies_count, 0) AS vacancies_count
FROM areas a
LEFT JOIN employers_count_per_area ec ON ec.area = a.id
LEFT JOIN vacancies_count_per_area vc ON vc.area_id = a.id
ORDER BY vacancies_count DESC, employers_count DESC;
"""

result_5_2 = pd.read_sql_query(query_5_2, connection)
print("5.2. Количество работодателей и вакансий по регионам:")
print(result_5_2.head(12))

# Найдем регион без вакансий с наибольшим количеством работодателей
regions_without_vacancies = result_5_2[result_5_2['vacancies_count'] == 0]
if len(regions_without_vacancies) > 0:
    max_employers_region = regions_without_vacancies.loc[regions_without_vacancies['employers_count'].idxmax()]
    print(f"\nРегион без вакансий с наибольшим количеством работодателей:")
    print(f"{max_employers_region['area']} - {max_employers_region['employers_count']} работодателей")
else:
    print("\nВсе регионы имеют вакансии")

C:\Users\akins\AppData\Local\Temp\ipykernel_13000\636095315.py:28: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_2 = pd.read_sql_query(query_5_2, connection)


5.2. Количество работодателей и вакансий по регионам:
               area  employers_count  vacancies_count
0            Москва             5864             5333
1   Санкт-Петербург             2217             2851
2             Минск             1115             2112
3       Новосибирск              573             2006
4            Алматы              721             1892
5      Екатеринбург              609             1698
6   Нижний Новгород              426             1670
7            Казань              480             1415
8         Краснодар              409             1301
9            Самара              315             1144
10   Ростов-на-Дону              335             1131
11          Воронеж              292             1063

Регион без вакансий с наибольшим количеством работодателей:
Россия - 410 работодателей


In [106]:
# Задание 5.3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Выберите максимальное значение из получившегося списка.
query_5_3 = """
SELECT 
    e.name as employer_name,
    COUNT(DISTINCT v.area_id) as regions_count
FROM employers e
LEFT JOIN vacancies v ON e.id = v.employer_id
GROUP BY e.id, e.name
ORDER BY regions_count DESC;
"""

result_5_3 = pd.read_sql_query(query_5_3, connection)
print("5.3. Количество регионов для каждого работодателя:")
print(result_5_3.head(10))
print(f"\nМаксимальное количество регионов: {result_5_3.iloc[0]['regions_count']}")
print(f"Работодатель с максимальным количеством регионов: {result_5_3.iloc[0]['employer_name']}")


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\46306375.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_3 = pd.read_sql_query(query_5_3, connection)


5.3. Количество регионов для каждого работодателя:
                employer_name  regions_count
0                      Яндекс            181
1                  Ростелеком            152
2                  Спецремонт            116
3      Поляков Денис Иванович             88
4                   ООО ЕФИН              71
5                  Совкомбанк             63
6                         МТС             55
7  ЭФКО, Управляющая компания             49
8                        КРОН             48
9                Почта России             48

Максимальное количество регионов: 181
Работодатель с максимальным количеством регионов: Яндекс


In [107]:
# Задание 5.4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.
query_5_4 = """
SELECT COUNT(*) as employers_without_industry
FROM employers e
LEFT JOIN employers_industries ei ON e.id = ei.employer_id
WHERE ei.employer_id IS NULL;
"""

result_5_4 = pd.read_sql_query(query_5_4, connection)
print("5.4. Количество работодателей без указанной сферы деятельности:")
print(result_5_4)
print(f"Ответ: {result_5_4.iloc[0]['employers_without_industry']} работодателей")


5.4. Количество работодателей без указанной сферы деятельности:
   employers_without_industry
0                        8419
Ответ: 8419 работодателей


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\255560940.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_4 = pd.read_sql_query(query_5_4, connection)


In [108]:
# Задание 5.5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.
query_5_5 = """
SELECT 
    e.name as company_name,
    COUNT(ei.industry_id) as industries_count
FROM employers e
INNER JOIN employers_industries ei ON e.id = ei.employer_id
GROUP BY e.id, e.name
HAVING COUNT(ei.industry_id) = 4
ORDER BY e.name
LIMIT 3;
"""

result_5_5 = pd.read_sql_query(query_5_5, connection)
print("5.5. Компании с четырьмя сферами деятельности (алфавитный порядок):")
print(result_5_5)
if len(result_5_5) >= 3:
    print(f"Ответ: {result_5_5.iloc[2]['company_name']}")
else:
    print("Недостаточно компаний с четырьмя сферами деятельности")


5.5. Компании с четырьмя сферами деятельности (алфавитный порядок):
   company_name  industries_count
0  101 Интернет                 4
1      21vek.by                 4
2          2ГИС                 4
Ответ: 2ГИС


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\1877182818.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_5 = pd.read_sql_query(query_5_5, connection)


In [109]:
# Задание 5.6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».
query_5_6 = """
SELECT COUNT(DISTINCT ei.employer_id) as employers_count
FROM employers_industries ei
INNER JOIN industries i ON ei.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения';
"""

result_5_6 = pd.read_sql_query(query_5_6, connection)
print("5.6. Количество работодателей в сфере 'Разработка программного обеспечения':")
print(result_5_6)
print(f"Ответ: {result_5_6.iloc[0]['employers_count']} работодателей")


5.6. Количество работодателей в сфере 'Разработка программного обеспечения':
   employers_count
0             3553
Ответ: 3553 работодателей


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\1815049731.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_6 = pd.read_sql_query(query_5_6, connection)


In [110]:
# Задание 5.7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании, собранных в этой таблице.
query_5_7 = """
WITH million_cities AS (
    SELECT name FROM areas 
    WHERE name IN (
        'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 
        'Нижний Новгород', 'Челябинск', 'Самара', 'Омск', 'Ростов-на-Дону',
        'Уфа', 'Красноярск', 'Воронеж', 'Краснодар', 'Пермь', 'Волгоград'
    )
),
yandex_vacancies AS (
    SELECT 
        a.name as area,
        COUNT(v.id) as vacancy_count
    FROM vacancies v
    JOIN employers e ON v.employer_id = e.id
    JOIN areas a ON v.area_id = a.id
    WHERE e.name = 'Яндекс' AND a.name IN (SELECT name FROM million_cities)
    GROUP BY a.name
),
vacancies_with_order AS (
    SELECT area, vacancy_count, 0 as is_total FROM yandex_vacancies
    UNION ALL
    SELECT 'Total' as area, SUM(vacancy_count), 1 as is_total FROM yandex_vacancies
)
SELECT area, vacancy_count
FROM vacancies_with_order
ORDER BY is_total, vacancy_count DESC;
"""

result_5_7 = pd.read_sql_query(query_5_7, connection)
print("5.7. Вакансии Яндекса в городах-миллионниках:")
print(result_5_7)
print(f"\nКоличество строк в выборке: {len(result_5_7)}")
print(f"Общее количество вакансий (Total): {result_5_7[result_5_7['area'] == 'Total']['vacancy_count'].iloc[0] if 'Total' in result_5_7['area'].values else 'Нет данных'}")


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\2243848697.py:31: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_5_7 = pd.read_sql_query(query_5_7, connection)


5.7. Вакансии Яндекса в городах-миллионниках:
               area  vacancy_count
0            Москва           54.0
1   Санкт-Петербург           42.0
2      Екатеринбург           39.0
3   Нижний Новгород           36.0
4       Новосибирск           35.0
5           Воронеж           32.0
6         Краснодар           30.0
7            Самара           26.0
8               Уфа           26.0
9            Казань           25.0
10            Пермь           25.0
11   Ростов-на-Дону           25.0
12        Волгоград           24.0
13       Красноярск           23.0
14        Челябинск           22.0
15             Омск           21.0
16            Total          485.0

Количество строк в выборке: 17
Общее количество вакансий (Total): 485.0


In [111]:
# Задание 6.1. Сколько вакансий имеет отношение к данным? Считаем, что вакансия имеет отношение к данным, если в ее названии содержатся слова 'data' или 'данн'.
query_6_1 = """
SELECT COUNT(*) as data_vacancies
FROM vacancies
WHERE LOWER(name) LIKE '%data%' OR LOWER(name) LIKE '%данн%';
"""

result_6_1 = pd.read_sql_query(query_6_1, connection)
print("6.1. Количество вакансий, связанных с данными:")
print(result_6_1)
print(f"Ответ: {result_6_1.iloc[0]['data_vacancies']} вакансий")


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\1153747207.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_1 = pd.read_sql_query(query_6_1, connection)


6.1. Количество вакансий, связанных с данными:
   data_vacancies
0            1771
Ответ: 1771 вакансий


In [112]:
# Задание 6.2. Сколько есть подходящих вакансий для начинающего дата-сайентиста?
query_6_2 = """
SELECT 
    COUNT(*) as junior_ds_vacancies
FROM vacancies
WHERE (
    LOWER(name) LIKE '%data scientist%' OR 
    LOWER(name) LIKE '%data science%' OR 
    LOWER(name) LIKE '%исследователь данных%' OR 
    (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%') OR 
    LOWER(name) LIKE '%machine learning%' OR 
    LOWER(name) LIKE '%машинн%обучен%'
) AND (
    LOWER(name) LIKE '%junior%' OR 
    LOWER(experience) = 'нет опыта' OR 
    LOWER(employment) = 'стажировка'
);
"""

result_6_2 = pd.read_sql_query(query_6_2, connection)
print("6.2. Количество вакансий для начинающего дата-сайентиста:")
print(result_6_2)
print(f"Ответ: {result_6_2.iloc[0]['junior_ds_vacancies']} вакансий")


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\4019921765.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_2 = pd.read_sql_query(query_6_2, connection)


6.2. Количество вакансий для начинающего дата-сайентиста:
   junior_ds_vacancies
0                   51
Ответ: 51 вакансий


In [113]:
# Задание 6.3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?
query_6_3 = """
SELECT COUNT(DISTINCT id) as ds_sql_vacancies
FROM vacancies
WHERE (
    LOWER(name) LIKE '%data scientist%' OR 
    LOWER(name) LIKE '%data science%' OR 
    LOWER(name) LIKE '%исследователь данных%' OR 
    (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%') OR 
    LOWER(name) LIKE '%machine learning%' OR 
    LOWER(name) LIKE '%машинн%обучен%'
) AND (
    LOWER(key_skills) LIKE '%sql%' OR 
    LOWER(key_skills) LIKE '%postgres%'
);
"""

result_6_3 = pd.read_sql_query(query_6_3, connection)
print("6.3. Количество вакансий для DS с навыками SQL или PostgreSQL:")
print(result_6_3)
print(f"Ответ: {result_6_3.iloc[0]['ds_sql_vacancies']} вакансий")


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\1294782708.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_3 = pd.read_sql_query(query_6_3, connection)


6.3. Количество вакансий для DS с навыками SQL или PostgreSQL:
   ds_sql_vacancies
0               229
Ответ: 229 вакансий


In [114]:
# Задание 6.4. С помощью запроса, аналогичного предыдущему, проверьте, насколько популярен Python в требованиях работодателей к DS. Вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.
query_6_4 = """
SELECT COUNT(DISTINCT id) as ds_python_vacancies
FROM vacancies
WHERE (
    LOWER(name) LIKE '%data scientist%' OR 
    LOWER(name) LIKE '%data science%' OR 
    LOWER(name) LIKE '%исследователь данных%' OR 
    (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%') OR
    LOWER(name) LIKE '%machine learning%' OR 
    LOWER(name) LIKE '%машинн%обучен%'
) AND (
    LOWER(key_skills) LIKE '%python%' 
);
"""

result_6_4 = pd.read_sql_query(query_6_4, connection)
print("6.4. Количество вакансий для DS с навыком Python:")
print(result_6_4)
print(f"Ответ: {result_6_4.iloc[0]['ds_python_vacancies']} вакансий")


6.4. Количество вакансий для DS с навыком Python:
   ds_python_vacancies
0                  357
Ответ: 357 вакансий


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\2616420102.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_4 = pd.read_sql_query(query_6_4, connection)


In [115]:
# Задание 6.5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
query_6_5 = """
SELECT 
    ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1)::numeric, 2) AS avg_skills_per_vacancy
FROM vacancies
WHERE (
    LOWER(name) LIKE '%data scientist%' OR 
    LOWER(name) LIKE '%data science%' OR 
    LOWER(name) LIKE '%исследователь данных%' OR 
    (name LIKE '%ML%' AND LOWER(name) NOT LIKE '%html%') OR 
    LOWER(name) LIKE '%machine learning%' OR 
    LOWER(name) LIKE '%машинн%обучен%'
);
"""

result_6_5 = pd.read_sql_query(query_6_5, connection)
print("6.5. Среднее количество ключевых навыков в вакансиях для DS:")
print(result_6_5)
print(f"Ответ: {result_6_5.iloc[0]['avg_skills_per_vacancy']} навыков в среднем")

C:\Users\akins\AppData\Local\Temp\ipykernel_13000\2861084122.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_5 = pd.read_sql_query(query_6_5, connection)


6.5. Среднее количество ключевых навыков в вакансиях для DS:
   avg_skills_per_vacancy
0                    6.41
Ответ: 6.41 навыков в среднем


In [116]:
# Задание 6.6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта (уникальное значение из поля experience).
query_6_6 = """
SELECT 
    experience, 
    round(AVG(coalesce((salary_from + salary_to) / 2, salary_from, salary_to)), 0) as avg_salary
FROM vacancies
WHERE (
    LOWER(name) LIKE '%data scientist%' OR 
    LOWER(name) LIKE '%data science%' OR 
    LOWER(name) LIKE '%исследователь данных%' OR 
    (name LIKE '%ML%' AND LOWER(name) NOT LIKE '%html%') OR 
    LOWER(name) LIKE '%machine learning%' OR 
    LOWER(name) LIKE '%машинн%обучен%'
) AND (
    salary_from IS NOT NULL OR salary_to IS NOT NULL
)
GROUP BY experience
ORDER BY avg_salary DESC;
"""

result_6_6 = pd.read_sql_query(query_6_6, connection)
print("6.6. Средняя зарплата для DS по опыту работы:")
print(result_6_6)

# Найдем зарплату для опыта от 3 до 6 лет
experience_3_6 = result_6_6[result_6_6['experience'] == 'От 3 до 6 лет']
if len(experience_3_6) > 0:
    print(f"\nСредняя зарплата для DS с опытом от 3 до 6 лет: {experience_3_6.iloc[0]['avg_salary']} руб.")
else:
    print("\nНет данных для опыта от 3 до 6 лет")

6.6. Средняя зарплата для DS по опыту работы:
           experience  avg_salary
0       От 3 до 6 лет    243115.0
1  От 1 года до 3 лет    139675.0
2           Нет опыта     74643.0

Средняя зарплата для DS с опытом от 3 до 6 лет: 243115.0 руб.


C:\Users\akins\AppData\Local\Temp\ipykernel_13000\1338599655.py:21: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_6_6 = pd.read_sql_query(query_6_6, connection)


# Выводы по блокам заданий

## Базовые характеристики базы данных (задания 3.1-3.4)
- **Общее количество вакансий**: Позволяет оценить масштаб рынка труда. Почти 50 тысяч вакансий, но этот показатель не дает информации о достаточности или избыточности вакансий на рынке труда, необходим комплексный анализ с учетом данных о резюме. По официальным данным в РФ безработных граждан в районе 1,7 млн. человек.
- **Количество работодателей**: Показывает разнообразие компаний на рынке (23,5 тысяч работодателей). В РФ на 1 сентября 2025 года в ЕГРЮЛ зарегестрировано 3,2 млн. организаций (без учета ИП), учитывая и общее количество безработных у Hh есть огромные перспективы для развития.
- **Количество регионов**: Отражает географическое покрытие вакансий (1,3 тысяч регионов). Данные содержат информацию не о регионах (в РФ 89 регионов), а о населенных пунктах (и не только российских), если учесть, что в РФ их в районе 155 тысяч, то можно предположить, что многие даже не знают что такое рынок труда в целом и Hh в частности
- **Количество сфер деятельности**: Демонстрирует разнообразие отраслей (294 отрасли)

## Географическое распределение (задания 4.1, 5.1-5.3, 5.7)
- **Лидеры по количеству вакансий**: Москва и Санкт-Петербург традиционно лидируют. Вакансии размещены только в 769 регионах (городах). На Москву приходится 10% от общего числа вакансий. 50% вакансий размещены в 12 городах.
- **Региональная специфика**: Некоторые регионы могут иметь высокую концентрацию работодателей при низком количестве вакансий. Только в Москве и Санкт-Петербурге на каждого работодателя приходится в среднем по одной вакансии. Далее в среднем на каждого работодателя приходится по 2 и более вакансий, например, в Воронеже на каждого из 292 работодателей приходится почти по 4 вакансии.
- **Компании-лидеры**: Крупные IT-компании (Яндекс) представлены в 181 регионе 1933 вакансии, в том числе в городах-миллионниках размещены 485 вакансий. На втором месте Ростелеком - в 152 регионах 492 вакансии.

## Анализ зарплат (задания 4.2-4.3, 6.6)
- **Заполненность зарплатных данных**: Показывает, какая доля вакансий содержит информацию о зарплате. Только 50% (24 тысячи вакансий) имеют информацию об уровне заработной платы.
- **Средние значения**: Дают представление о рыночном уровне оплаты труда. Работодатели в среднем готовы платить от 71 до 111 тыс.руб.
- **Специализация по опыту**: Зарплаты Data Science специалистов сильно зависят от опыта работы. Уровень зарплаты специалистов данной сферы выше среднего уровня и с учетом опыта работы составляют от 75 до 243 тыс.руб.

## Требования к кандидатам (задания 4.4-4.5, 6.1-6.5)
- **Гибкость трудоустройства**: Сочетание графика работы и типа трудоустройства. Больше всего 35 тысяч вакансий требуется работников на полный рабочий день и полную занятость, на втором месте это удаленная работа на полную занятость 8 тысяч вакансий. Самые непопулярные у работодателей сочетания вахтового и сменного графика с проектной работой.
- **Опыт работы**: Распределение вакансий по требованиям к опыту. Большинство вакансий (26 тысяч) содержат требования к опыту работы от 1 года до 3 лет. 30% от общего числа вакансий (15 тысяч) для работников с опытом работы от 3 до 6 лет. Кроме того работодатели охотнее ищут соискателей без опыта работы, чем работников с опытом более 6 лет.
- **Ключевые навыки**: Популярность SQL, Python и других технологий в Data Science, в среднем такие специалисты должны иметь не менее 6 ключевых навыков. 1771 вакансия связаны с данными и их обработкой, в том числе 51 вакансия для начинающих дата-сайентистов.

## Сферы деятельности (задания 5.4-5.6)
- **IT-доминирование**: "Разработка программного обеспечения" - одна из самых популярных сфер (15% работодателей указали данную сферу деятельности)
- **Многопрофильность**: Некоторые компании работают в нескольких сферах одновременно, максимальное количество сфер деятельности для одного работодателя в представленных в данных 5.
- **Неполнота данных**: Часть работодателей не указывает сферу деятельности (8 тысяч работодателей)


# Общий вывод по результатам анализа вакансий

## Каких вакансий много?
1. **IT-вакансии доминируют**: Особенно в сфере разработки программного обеспечения
2. **Data Science набирает популярность**: Растет количество вакансий, связанных с анализом данных и машинным обучением
3. **Москва и СПб лидируют**: Столичные регионы концентрируют основную массу вакансий
4. **Средний опыт работы**: Большинство вакансий требуют опыт от 1 до 3 лет

## Рынок труда в различных городах
1. **Централизация**: Москва и Санкт-Петербург концентрируют основную часть вакансий
2. **Региональные особенности**: Некоторые регионы имеют специфические отрасли
3. **Удаленная работа**: Возможность работы из разных регионов расширяет географию поиска
4. **Зарплатные различия**: Столичные зарплаты значительно выше региональных

## Тенденции и прогнозы
1. **Рост Data Science**: Продолжающийся рост спроса на специалистов по данным
2. **Гибкость работы**: Увеличение доли удаленной работы и гибких графиков
3. **Технологический стек**: Python и SQL остаются ключевыми навыками
4. **Опыт vs навыки**: Баланс между требованиями к опыту и техническим навыкам


In [117]:
# Закрытие соединения с базой данных
connection.close()
print("Соединение с базой данных закрыто.")


Соединение с базой данных закрыто.
